Prometheus support, Part 1

PhanLe1010 opened the pull request:
Added the following metrics:
1. longhorn_volume_capacity_bytes
1. longhorn_volume_actual_size_bytes
1. longhorn_node_stats_node_status

longhorn/longhorn#1180

joshimoo commented on api/router.go:
```diff
@@ -1,6 +1,7 @@
 package api
 
 import (
+	"github.com/longhorn/longhorn-manager/monitoring"
```
add new section after `apierrors "k8s.io/apimachinery/pkg/api/errors"` a new line acts as a separator between the sections.

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/backup_stats_number_failed_backups_metric.go:
```diff
@@ -0,0 +1,47 @@
+		return err
+	}
+
+	for _, v := range volumeLists.Items {
```
Don't store these persistently, in the objects, this will create a lot of update and writes which are completely unnecessary. 

Metrics should only exist at runtime in memory => hashmap
Prometheus can collect these, from your metrics endpoints.

joshimoo replied:
applies to all the other metrics as well, also see implementation comment above for the increment methods etc.

PhanLe1010 replied:
We will discuss more about this on the meeting tommorow

joshimoo commented on manager/volume.go:
```diff
@@ -741,3 +741,39 @@ func (m *VolumeManager) UpdateDataLocality(name string, dataLocality types.DataL
 	return v, nil
 }
+
+func (m *VolumeManager) IncBackupFailedCount(name string) (v *longhorn.Volume, err error) {
```
Remove these methods and add a metrics interface, then have metric class implementations in a datastore class.

example:
metrics.backup.succeeded(volumeName)
metrics.backup.failed(volumeName)

metrics is a singleton instance that contains references to the actual metric interface implementation.
this looks nice and makes it easy to split the metrics across different files.

Don't store these persistently, in the objects, this will create a lot of update and writes which are completely unnecessary. See below on how to store these in memory.

Also don't return any errors from these methods, metrics are best effort only.

PhanLe1010 replied:
Follow your suggestion, I created a new collector for each type (volumeCollector, backupCollector, nodeCollector, etc..) and have a common baseCollector. 

> Don't store these persistently, in the objects, this will create a lot of update and writes which are completely unnecessary. See below on how to store these in memory.

We will discuss more about this on the meeting tommorow

joshimoo commented on monitoring/backup_stats_number_succeed_backups_metric.go:
```diff
@@ -0,0 +1,47 @@
+package monitoring
+
+import (
+	"github.com/pkg/errors"
```
add different sections for import

PhanLe1010 replied:
Thank you! This file was deleted

joshimoo commented on monitoring/collector.go:
```diff
@@ -0,0 +1,105 @@
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/pkg/errors"
```
add different sections for import

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/collector.go:
```diff
@@ -0,0 +1,105 @@
+	namespace     string
+	currentNodeID string
+
+	log *logrus.Logger
```
Use logger as name to be consistent with the different controllers.
use logrus.FieldLogger as type, and pass in the base logger in the new NewCollector() call
see format of the different controllers as example.

Then you can add a method, if you need per metric logging with different fields for example.
log := getLoggerForMetrics(collector.logger, metric)

PhanLe1010 replied:
Thank you! Fixed to use logging structure

joshimoo commented on monitoring/monitoring.go:
```diff
@@ -0,0 +1,22 @@
+package monitoring
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
```
add different sections.

PhanLe1010 replied:
Thank you! Fixed

joshimoo reviewed (changes requested):
Will review the rest after you had a chance to address this feedback.

PhanLe1010 commented on api/snapshot.go:
```diff
@@ -143,7 +143,11 @@ func (s *Server) SnapshotRevert(w http.ResponseWriter, req *http.Request) (err e
+	var volName string
 	defer func() {
+		if err != nil && volName != "" {
+			_, err = s.m.IncBackupFailedCount(volName)
```
Thank you! Fix so that we only count at at the manager level

PhanLe1010 commented on monitoring/backup_stats_number_failed_backups_metric.go:
```diff
@@ -0,0 +1,47 @@
+package monitoring
+
+import (
+	"github.com/pkg/errors"
```
Thank you! This file was deleted

PhanLe1010 commented:
Thank you Joshua! I have fixed 4/5 suggestions.

Things that I fixed:
1. logging structure
1. use DS instead of calling etcd API directly in all metrics
1. Create a new collector for each type (volumeCollector, backupCollector, nodeCollector, etc..) and have a common baseCollector. This structure is similar to controllers. We have volumeController, nodeController, etc.. which have a commend baseController
1. rewrite the nodeStatus metrics so that it use 1 more dimension for different type of conditions. e.g longhorn_node_status{node_id="node1", condition="ready"}, longhorn_node_status{node_id="node1", condition="schedulabe"}, etc..

One thing that I would like to discuss more:
The best way to export information about backups. Option 1: expose the number of failed backup and succeed backups for each volume. Option 2: expose backup status for each backup as Portworx does: https://docs.portworx.com/reference/metrics/ 

- If we chose Option 1 above, what is the best way to present the metric to users. And what would be the best way to persist the data across diffrent scrapes.
- Option 2 seems to be very helpful from users' perspective but hard to implement with the current Longhorn's implementation

joshimoo commented on monitoring/volume_collector.go:
```diff
@@ -0,0 +1,64 @@
+func (vc *VolumeCollector) Collect(ch chan<- prometheus.Metric) {
+	volumeLists, err := vc.ds.ListVolumesRO()
+	if err != nil {
+		vc.logger.Errorf("error during scrape %v", err)
```
WithError(err)

PhanLe1010 replied:
Thank you! Fixed

joshimoo reviewed (commented):
We can discuss the backup stuff in zoom later.

PhanLe1010 commented on monitoring/node_stats_node_status_metric.go:
```diff
@@ -0,0 +1,76 @@
+package monitoring
+
+import (
+	"github.com/longhorn/longhorn-manager/types"
```
This is outdate. This file was deleted

PhanLe1010 commented on monitoring/node_stats_node_status_metric.go:
```diff
@@ -0,0 +1,76 @@
+		err = errors.Wrap(err, "failed to get values for nodeStatusMetric")
+	}()
+
+	nodeLists, err := c.lhClient.LonghornV1beta1().Nodes(c.namespace).List(metav1.ListOptions{
```
This is outdate. This file was deleted

PhanLe1010 commented on monitoring/monitoring.go:
```diff
@@ -0,0 +1,37 @@
+	return promhttp.HandlerFor(longhornCustomRegistry, promhttp.HandlerOpts{})
+}
+
+func InitMonitoringSystem(logger logrus.FieldLogger, ds *datastore.DataStore, currentNodeID string) {
```
Thank you for sharing! It is awsome!

PhanLe1010 commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,76 @@
+func (nc *NodeCollector) Collect(ch chan<- prometheus.Metric) {
+	nodeList, err := nc.ds.ListNodesRO()
+	if err != nil {
+		nc.logger.Errorf("error during scrape %v", err)
```
Thank you! I fixed it

PhanLe1010 commented on datastore/longhorn.go:
```diff
@@ -1032,6 +1032,11 @@ func (s *DataStore) ListNodes() (map[string]*longhorn.Node, error) {
 	return itemMap, nil
 }
 
+// ListNodesRO returns a list of all Nodes for the given namespace
```
Thank you for sharing! I fixed it

PhanLe1010 commented:
Removed the backup_collector file. This PR only contains volume capacity, volume actual size, and node status metrics.

PhanLe1010 commented:
@joshimoo Can you please take a look at this PR again?

joshimoo commented on monitoring/types.go:
```diff
@@ -0,0 +1,22 @@
+	subsystemBackupStats = "backup_stats"
+	subsystemNodeStats   = "node_stats"
+
+	prometheusNodeIDLabel    = "node_id"
```
remove the `id` suffix and just use `node, volume, condition, condition_reason`

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/types.go:
```diff
@@ -0,0 +1,22 @@
+
+	subsystemVolume      = "volume"
+	subsystemBackupStats = "backup_stats"
+	subsystemNodeStats   = "node_stats"
```
remove the `stats` suffix and just use `node`
this way it will look nicer and be easier to read:

`node_status` vs `node_stats_node_status`

this also makes sense since all our per component collectors are individual metrics and not aggregates.
If we want to add different aggregated statistics we could add a collector like this instead:
`stats_nodes`
`stats_volumes`

But at the moment I don't see any direct need for this, since once all individual metrics are tracked via prometheus the user can use the prometheus query language to generate the aggregates themselves.
Which is much better then us generating them.

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/types.go:
```diff
@@ -0,0 +1,22 @@
+	longhornName = "longhorn"
+
+	subsystemVolume      = "volume"
+	subsystemBackupStats = "backup_stats"
```
remove the backup related field for now, since we are not doing it till later.

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,77 @@
+
+		for _, conditionType := range []string{types.NodeConditionTypeReady, types.NodeConditionTypeSchedulable, types.NodeConditionTypeMountPropagation} {
+			val := 0
+			switch condition := types.GetCondition(node.Status.Conditions, conditionType); condition.Status {
```
Use binary conditions:
False/Unkown => 0
True => 1

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,77 @@
+			case types.ConditionStatusUnknown:
+				val = 3
+			}
+			ch <- prometheus.MustNewConstMetric(nc.newNodeStatusMetric.Desc, nc.newNodeStatusMetric.Type, float64(val), nc.currentNodeID, strings.ToLower(conditionType))
```
Add the condition_reason values as a dimension.
This allows for example to find how often a node became not ready because of network failure, cpu failure, etc.

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,77 @@
+			continue
+		}
+
+		for _, conditionType := range []string{types.NodeConditionTypeReady, types.NodeConditionTypeSchedulable, types.NodeConditionTypeMountPropagation} {
```
Instead of checking for specific conditions, return all node conditions, by iterating over the node.Status.Conditions map.

joshimoo replied:
The same approach can be used for Volume Conditions, Disk Conditions, etc down the line.

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,77 @@
+
+type NodeCollector struct {
+	*baseCollector
+	newNodeStatusMetric metricInfo
```
Why is this called `newNodeStatusMetric`?
could it be called just `statusMetric` since it's private and inside of the NodeCollector?

PhanLe1010 replied:
Opps. this is a typo. Thank you! Fixed

joshimoo commented on monitoring/volume_collector.go:
```diff
@@ -0,0 +1,65 @@
+
+type VolumeCollector struct {
+	*baseCollector
+	volumeCapacityMetric   metricInfo
```
maybe just `capacityMetric`?

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/volume_collector.go:
```diff
@@ -0,0 +1,65 @@
+type VolumeCollector struct {
+	*baseCollector
+	volumeCapacityMetric   metricInfo
+	volumeActualSizeMetric metricInfo
```
maybe just `sizeMetric`?

PhanLe1010 replied:
Thank you! Fixed

joshimoo commented on monitoring/volume_collector.go:
```diff
@@ -0,0 +1,65 @@
+
+	vc.volumeActualSizeMetric = metricInfo{
+		Desc: prometheus.NewDesc(
+			prometheus.BuildFQName(longhornName, subsystemVolume, "actual_size_bytes"),
```
maybe just `used_bytes` ? 

This can be used to get the ratio of used / capacity and do alerting in the case when there are to many snapshots for a volume for example. (i.e. once ratio > 5 for example)

joshimoo replied:
I see elsewhere you use the term `usage` instead, so let's stick with that?

PhanLe1010 replied:
Thank you! Fixed

joshimoo reviewed (commented):
as discussed in zoom, looks good to me in general, some smaller naming suggestions below.

joshimoo commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,72 @@
+			if condition.Status == types.ConditionStatusTrue {
+				val = 1
+			}
+			ch <- prometheus.MustNewConstMetric(nc.statusMetric.Desc, nc.statusMetric.Type, float64(val), nc.currentNodeID, strings.ToLower(condition.Type), condition.Reason)
```
As an alternative you can also just use
`defer func() { recover() }` at the top of the collectNodeStatus function.
This will swallow the panic, and stop execution of the collectNodeStatus function.

In there you could then log once,  instead of per node / condition / etc.
This is probably a simpler design, since generally these shouldn't panic anyway.

PhanLe1010 replied:
Yeah!!! Thank you! Fixed on the https://github.com/longhorn/longhorn-manager/pull/684 PR

joshimoo commented on monitoring/volume_collector.go:
```diff
@@ -0,0 +1,65 @@
+	}
+	for _, v := range volumeLists {
+		if v.Status.OwnerID == vc.currentNodeID {
+			ch <- prometheus.MustNewConstMetric(vc.capacityMetric.Desc, vc.capacityMetric.Type, float64(v.Spec.Size), vc.currentNodeID, v.Name)
```
see comment above about panicking, I think using the `defer func { recover() }` approach is okay for the collection methods.

PhanLe1010 replied:
Thank you! Fixed on the https://github.com/longhorn/longhorn-manager/pull/684 PR

joshimoo reviewed (changes requested):
Added concerns about the panicking behavior.

PhanLe1010 commented on monitoring/node_collector.go:
```diff
@@ -0,0 +1,72 @@
+	ds *datastore.DataStore) *NodeCollector {
+
+	nc := &NodeCollector{
+		baseCollector: newBaseCollector("node", logger, nodeID, ds),
```
Thank you! Fixed on the https://github.com/longhorn/longhorn-manager/pull/684 PR

PhanLe1010 commented on monitoring/monitoring.go:
```diff
@@ -0,0 +1,35 @@
+// MustRegister registers the provided Collectors with the longhornCustomRegistry and
+// panics if any error occurs.
+func MustRegister(collector prometheus.Collector) {
+	longhornCustomRegistry.MustRegister(collector)
```
Thank you! Fixed on the https://github.com/longhorn/longhorn-manager/pull/684 PR

joshimoo commented:
superseded by PR #684
